lua/telescope/actions/history.lua

```diff
@@ -0,0 +1,218 @@
+      assert(not write_err, write_err)
+      uv.fs_close(fd, function(close_err)
+        assert(not close_err, close_err)
+      end)
+    end)
+  end)
+end
+
+local append_async = function(path, txt) write_async(path, txt, "a") end
+
+local m = {}
+
+local SimpleHistory = {}
+SimpleHistory.__index = SimpleHistory
+
+function SimpleHistory:new()
+  local obj = {
+    enabled = true
+  }
+  if not conf.history_location then
+    obj.enabled = false
+    return setmetatable(obj, self)
+  end
+  if conf.history_limit then
+    obj.limit = conf.history_limit
+  end
+
+  obj.path = vim.fn.expand(conf.history_location)
+  local p = Path:new(obj.path)
+  if not p:exists() then p:touch({ parents = true }) end
+
+  obj.content = Path:new(obj.path):readlines()
+  obj.index = #obj.content
+  table.remove(obj.content, obj.index)
+
+  return setmetatable(obj, self)
+end
+
+function SimpleHistory:reset()
+  if not self.enabled then return end
+  self.index = #self.content + 1
+end
+
+function SimpleHistory:append(line)
+  if not self.enabled then return end
+
+  if line ~= '' then
+    if self.content[#self.content] ~= line then
+      table.insert(self.content, line)
+
+      local len = #self.content
+      if self.limit and len > self.limit then
+        -- We need to remove the first diff items not only the first one
+        -- because a history can already be like 60 items but the user changes it
+        -- to 50. Now we need to remove the no longer needed items.
+        -- We could do this check on startup as well but currently i don't think its
+        -- necessary.
+        local diff = len - self.limit
+        for i = diff, 1, -1 do
+          table.remove(self.content, i)
+        end
+        write_async(self.path, table.concat(self.content, '\n') .. '\n', 'w')
+      else
+        append_async(self.path, line .. '\n')
+      end
+    end
+  end
+  self:reset()
+end
+
+function SimpleHistory:get_next()
+  if not self.enabled then return end
+
+  local next_idx = self.index + 1
+  if next_idx <= #self.content then
+    self.index = next_idx
+    return self.content[next_idx]
+  end
+  self.index = #self.content + 1
+  return nil
+end
+
+function SimpleHistory:get_prev()
+  if not self.enabled then return end
+
+  local next_idx = self.index - 1
+  if next_idx >= 1 then
+    self.index = next_idx
+    return self.content[next_idx]
+  end
+  return nil
+end
+
+m.get_simple_history = function(...)
+  return SimpleHistory:new(...)
+end
+
+------------------------------------------------------------------------------
+------------------------------------------------------------------------------
+-- DB HISTORY SCRATCH. MOVING TO EXTENSION AFTER MAKING IT WORK --------------
+
+local sql = require('sql')
+
+-- TODO(conni2461): Serialize metadata
+-- TODO(conni2461): CURRENTLY HARDCODED picker and cwd
+local SmartHistory = {}
+SmartHistory.__index = SmartHistory
+
+function SmartHistory:new()
+  local obj = {
+    enabled = true
+  }
```

Comment:
you can make the obj to = `(sql.new(path)):table('histories', {})`

this way we can use self:insert, self:get ... etc, u can also add additional methods or overwrite existing methods id you like . see https://github.com/nvim-telescope/telescope-cheat.nvim/blob/master/lua/telescope/_extensions/cheat/db.lua

Reply:
This enable is for the history feature. No path not enabled.

Reply:
oh you can add it like
`obj.smart_enabled ` or enabled to the table object itself. The idea here is to reuse sql table object without need to prefix it as data, self = smarthistory. So in other functions you can do self: to access both your object and sql table object methods and variables

Reply:
I think you don't get the whole architecture. I don't get what you want from me :laughing: You can still access the db with this, but you shouldn't because you don't know whats behind the table. It could be DBHistory or it could be SimpleHistory. So i kinda want to keep it the same. So if someone wants to do something outside of History, he can get it with:

```lua
local history = action_state.get_current_history()
history.enabled -- will now show if we have a history or not, works for both
history.content -- get content, need to rename the object my bad
history:get_next() -- 
```

Reply:
Oh I see what you mean. Yah merging the two is bad idea in case of switching things around with something that implement get/insert method. Good call